content/news/2023-03-04-bevy-0.10/index.md

```diff
@@ -686,6 +686,87 @@ app from the app to send it between the main thread and the rendering thread. Th
 only valid to do after all the plugin build methods have been called, because any plugin may
 want to modify the rendering sub app.
 
+## ShaderDef Values
+
+<div class="release-feature-authors">authors: @mockersf</div>
+
+Bevy's shader processor now supports ShaderDefs with values. This allows developers to pass constant values into their shaders:
+
+```rust
+let shader_defs = vec![
+    ShaderDefVal::Int("MAX_DIRECTIONAL_LIGHTS".to_string(), 10),
+];
+```
+
+These can be used in `#if` statements to selectively enable shader code based on the value:
+
+```rust
+#if MAX_DIRECTIONAL_LIGHTS >= 10
+let color = vec4<f32>(1.0, 0.0, 0.0, 1.0);
+#else
+let color = vec4<f32>(0.0, 1.0, 0.0, 1.0);
+#endif
+```
+
+ShaderDef values can be inlined into shaders:
+
+```rust
+for (var i: u32 = 0u; i < #{MAX_DIRECTIONAL_LIGHTS}; i = i + 1u) {
+}
+```
+
+They can also be defined inline in shaders:
+
+```rust
+#define MAX_DIRECTIONAL_LIGHTS 10 
```

Comment:
Here the 10 is not clear to me, is this a default value?

Reply:
Its just an arbitrary value chosen to illustrate what setting values look like. Its equivalent to `let max_directional_lights = 10` in rust.

Reply:
I get that. What I don't understand is how this fits in with passing constant values in from rust. If I passed in a value of 15 using `ShaderDefVal::Int("MAX_DIRECTIONAL_LIGHTS".to_string(), 15)` would that override the value defined in the shader?

Reply:
Ahh yeah the precedence order is worth clarifying. Good call!